File: factory/src/main/java/my/sample/factory/impl/DefaultSampleFactoryService.java
package my.sample.factory.impl;

import my.sample.activator.SampleLogger;
import my.sample.factory.SampleFactoryService;

import java.util.HashMap;
import java.util.Map;

/**
 * @author mohammad shamsi <m.h.shams@gmail.com>
 */
public class DefaultSampleFactoryService implements SampleFactoryService {
    private SampleLogger logger;
    private Map<String, Object> config;

    public DefaultSampleFactoryService() {
        this.config = new HashMap<>();
    }

    public void init() {
        logger.info("Initializing DefaultSampleFactoryService: " + this);
    }

    public void update(Map<String, ?> config) {
        this.config.clear();
        this.config.putAll(config);

        logger.info("updating....: " + this);
    }

    public void destroy() {
        logger.info("Destroying DefaultSampleFactoryService: " + this);
    }

    public String echo(String message) {
        return "Echo processed: " + message;
    }

    @SuppressWarnings("UnusedDeclaration")
    public void setLogger(SampleLogger logger) {
        this.logger = logger;
    }

    @Override
    public String toString() {
        StringBuilder builder = new StringBuilder("{");
        for (String key : config.keySet()) {
            builder.append("\n\t").append(key).append(": ").append(config.get(key));
        }
        builder.append("\n").append("}");

        return builder.toString();
    }
}

File: factory/src/main/kotlin/my/sample/factory/impl/DefaultSampleFactoryService.kt
package my.sample.factory.impl

import my.sample.factory.SampleFactoryService

/**
 * @author mohammad shamsi <m.h.shams@gmail.com>
 */
public class DefaultSampleFactoryService : SampleFactoryService {
    private var foo: String? = null
    private var bar: String? = null
    private var baz: String? = null

    {
        println("new DefaultSampleFactoryService: $this")
    }

    public fun init() {
        println("Initializing DefaultSampleFactoryService: $this")
    }

    public fun update(config: Map<String, *>) {
        println("updating....: $this")
    }

    public fun destroy() {
        println("Destroying DefaultSampleFactoryService: $this")
    }

    override fun echo(message: String): String {
        return "Echo processed: $message"
    }

    public fun setFoo(foo: String) {
        println("setting new foo: $foo")
        this.foo = foo
    }

    public fun setBar(bar: String) {
        println("setting new bar: $bar")
        this.bar = bar
    }

    public fun setBaz(baz: String) {
        println("setting new baz: $baz")
        this.baz = baz
    }

    override public fun toString() = " InternalBean { foo = '$foo' , bar = '$bar', baz = '$baz' }"
}